### PR TITLE
feat: support TypeScript 5.5

### LINKED ISSUE #8996: TypeScript 5.5 Support
https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/

Please read through our TypeScript version policy: https://typescript-eslint.io/users/dependency-versions#typescript

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

_Please be patient._

------

# 🔲 Stricter Parsing for Decorators

We'll need to double check we correctly define our AST types and have an error in polace for this.

# 🔲 lib.d.ts Updates

We will need to regenerate our types within scope-manager.

# Other changes with no impact to us 

- Inferred Type Predicates
    - Shower thought: maybe there's space for a rule to warn against an annotation that would clobber the inferred predicate...?
- Control Flow Narrowing for Constant Indexed Accesses
- The JSDoc `@import` Tag
    - We don't support JSDoc at all so no action for us
- Regular Expression Syntax Checking
- Support for New ECMAScript `Set` Methods
- Isolated Declarations
    - Shower thought: We could probably consider killing `explicit-module-boundary-types` now considering they are both enforcing the same thing really.
- The `${configDir}` Template Variable for Configuration Files
    - This shouldn't have any impact for us, I think, cos it should just magically be handled by TS when we ask it for a parsed tsconfig.
- Consulting `package.json` Dependencies for Declaration File Generation
- Editor and Watch-Mode Reliability Improvements
- Performance and Size Optimizations
    - We'll benefit from these improvements though!
- Easier API Consumption from ECMAScript Modules
- The `transpileDeclaration` API
- Notable Behavioral Changes
	- Disabling Features Deprecated in TypeScript 5.0
	- `undefined` is No Longer a Definable Type Name
        - We don't concern ourselves with this level of error directly
    - Simplified Reference Directive Declaration Emit

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8996, fixes $7534
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Pretty straightforward. No changes impacted us aside from API deprecations.

### CODE REVIEW

**Review by JoshuaKGoldberg — APPROVED** [review]
South Park's Eric Cartman saying "SWEET"

**Review by kirkwaiblinger — APPROVED** [review]
🚀

File: packages/eslint-plugin/src/rules/prefer-includes.ts [hunk 1]
```diff
@@ -119,9 +119,7 @@ export default createRule({
       }
 
       // To string.
-      return String.fromCodePoint(
-        ...chars.map(c => (c as RegExpAST.Character).value),
-      );
+      return String.fromCodePoint(...chars.map(c => c.value));
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-includes.ts:121]: _swell_

**Comment by RobinTail** [comment]
@JoshuaKGoldberg , could you please also release it in the v8, the version that supports ESLint 9, please?

**Comment by bradzacher** [comment]
We will merge main into v8 when we have the opportunity. Please be patient.

**Pull request merged**